Fix failing log_artifact due to existing directory on FTP server

kafendt opened the pull request:
## What changes are proposed in this pull request?
 
When running log_artifact() twice without adding
an artifact_path the upload fails due to ftp.mkd(artifact_dir)
failing. ftp.mkd(artifact_dir) fails because artifact_dir already exists on
the server after the first call. This is now avoided by first checking whether the directory
already exists on the server before trying to create it.
 
## How is this patch tested?

Tested manually with a vsftpd ftp, I think the existing unittests should suffice.
 
## Release Notes
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Fixes bug which caused ftp artifact upload to fail on subsequent calls to log_artifact due to already existing directory.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [x] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Commit:
Fix failing log_artifact due to existing directory

When running log_artifact() twice without adding
an artifact_path the upload fails due to ftp.mkd(artifact_dir)
failing. This is now avoided by first checking whether the directory
already exists on the server.

aarondav commented on mlflow/store/ftp_artifact_repo.py:
```diff
@@ -50,7 +50,9 @@ def _is_dir(self, full_file_path):
             try:
-                ftp.mkd(artifact_dir)
+                head, _ = posixpath.split(artifact_dir)
+                if artifact_dir not in ftp.nlst(head):
```
I couldn't find a super obvious API, but is there no standard way to stat for existence of a specific file, as opposed to listing all files in the directory?

kafendt replied:
I also wondered about that but as it seems ftplib (or ftp itself) doesn't provide a way to check for existence of a folder. Apparently using `nlst` is the only way to realize this.

kafendt replied:
I actually just realized that there is a `_is_dir` function, shame on me. Now using that one :smile:

aarondav commented:
Perhaps we could add a unit test that mocks the multiple log artifacts call and that ensures we only make the subdirectory once? Of course, we have to mock the interaction, but at least it gives us some coverage of this situation.

kafendt committed:
use self._is_dir instead of ftp.nlst for check

also adjust the tests as they now call ftp.cwd twice
every time log_artifact is called

kafendt committed:
Add test for calling log_artifact multiple times

- tests simple logging of artifact
- tests logging another artifact in the same folder
- tests logging an artifact with artifact_path set

kafendt commented:
> Perhaps we could add a unit test that mocks the multiple log artifacts call and that ensures we only make the subdirectory once? Of course, we have to mock the interaction, but at least it gives us some coverage of this situation.

Test has been added

aarondav commented:
LGTM - thanks!
